keystore/v2/keystore/filesystem/backend/filesystem.go

```diff
@@ -150,6 +166,63 @@ func OpenDirectoryBackend(root string) (*DirectoryBackend, error) {
 	}, nil
 }
 
+func versionFilePath(rootDir string) string {
+	return filepath.Join(rootDir, versionFile)
+}
+
+// CheckDirectoryVersion checks whether a key directory is of expected version.
+func CheckDirectoryVersion(rootDir string) error {
+	content, err := ioutil.ReadFile(versionFilePath(rootDir))
+	if err != nil {
+		return err
+	}
+	if string(content) != versionString {
+		return ErrInvalidVersion
+	}
+	return nil
+}
+
+func checkVersionFile(rootDir string) error {
+	// First, check whether we already have a valid version file. If so then we're done.
+	// Otherwise, create a new version file if and only if it does not exist yet.
+	// It might also be that the file has been removed, but we can't tell that for sure,
+	// so just reinitialize the file in that case.
+	err := CheckDirectoryVersion(rootDir)
+	if err == nil {
+		return nil
+	}
+	if !os.IsNotExist(err) {
+		return err
+	}
+	return createVersionFile(rootDir)
+}
+
+func createVersionFile(rootDir string) error {
+	path := versionFilePath(rootDir)
+	// Make sure the file does not exist and create it with proper mode.
+	file, err := os.OpenFile(path, os.O_CREATE|os.O_EXCL|os.O_WRONLY, versionPerm)
+	if err != nil {
+		return err
+	}
+	// Close() might fail for newly written files, make sure we don't lose this error.
```

Comment:
did you check that it works like this? https://play.golang.org/p/yNPWMyA8TAt - checked, it works only if we re-assign value for return variable

Reply:
Hm... you’re right. It does not work as expected. (And I did not test it properly, shame on me.)

I saw this pattern somewhere on the web, or reading through the standard library, or something. Maybe it *did* use return variables there.

Thank you for attention and correction. I’ll review defers to see if this is incorrectly used anywhere else...

Reply:
I've looked through the code and it seems this is the only place which is incorrect. Fun thing, I *did* use this pattern correctly elsewhere:

- https://github.com/cossacklabs/acra/blob/e22cacaf4f03a9ec475fee76af330384556fb1d2/keystore/v2/keystore/filesystem/keyStoreLoad.go#L30-L52

I guess it's just a brain fart this time.

I've fixed the code to use return variables.